File: lintcode/number_of_islands.py
class Solution:
    # @param {boolean[][]} grid a boolean 2D matrix
    # @return {int} an integer
    def numIslands(self, grid):
        def helper(row, col):
            if 0 <= row < len(grid) and 0 <= col < len(grid[row]) \
                and grid[row][col] == 1:
                grid[row][col] = 0
                map(helper, (row + 1, row - 1,     row,     row),
                            (    col,     col, col + 1, col - 1))
                return 1
            return 0
        return sum(helper(row, col) for row in xrange(len(grid))
                                    for col in xrange(len(grid[row])))

File: leetcode/number_of_islands.js
/* Given a 2d grid map of '1's (land) and '0's (water), count the number of 
 * islands. An island is surrounded by water and is formed by connecting 
 * adjacent lands horizontally or vertically. You may assume all four edges of 
 * the grid are all surrounded by water.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    function helper(row, col) {
        if (0 <= row && row < grid.length
         && 0 <= col && grid[row].length
         && grid[row][col] === '1') {
            grid[row][col] = '0'
            helper(row + 1, col)
            helper(row - 1, col)
            helper(row, col + 1)
            helper(row, col - 1)
            return 1
        }
        return 0
    }
    for (var result = 0, i = 0; i < grid.length; ++i)
        for (var j = 0; j < grid[i].length; ++j)
            result += helper(i, j)
    return result
};
